### PR TITLE
FC to Conv2D

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+struct ReplaceFCWithConv2DPattern
+    : public OpRewritePattern<TFL::FullyConnectedOp> {
+  using OpRewritePattern<TFL::FullyConnectedOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TFL::FullyConnectedOp fcOp,
+                                PatternRewriter &rewriter) const override {
+    // Check for invalid types and return
+    // Input type must be QI8
+    if (!(fcOp.input()
```
Comment (andrewstanfordjason) [inline experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: I feel like we should pull this into a header after this merge as many will start with this.

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                  .cast<quant::QuantizedType>()
+                  .getStorageTypeIntegralWidth() == 8)) {
+      return failure();
+    }
+
+    // Filter type must be QI8
+    if (!(fcOp.filter()
+              .getType()
+              .cast<ShapedType>()
+              .getElementType()
```
Comment (andrewstanfordjason) [inline experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: I would hoist 
               fcOp.filter()
               .getType()
               .cast<ShapedType>()
               .getElementType()

into its own variable(same for the others). I find that much more readable.

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                  .getType()
+                  .cast<ShapedType>()
+                  .getElementType()
+                  .cast<quant::QuantizedType>()
+                  .getStorageTypeIntegralWidth() == 32)) {
+      return failure();
+    }
+
+    if (fcOp.weights_format() != "DEFAULT") {
+      return failure();
```
Comment (andrewstanfordjason) [inline experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp:69]: could we return an error via  fcOp.emitOpError() << "stuff"?

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        rewriter.create<TFL::ReshapeOp>(fcOp.getLoc(), expandedInputResultType,
+                                        fcOp.input(), expandedShapeConstantOp);
+
+    // Expand filter to 4 dims
+    auto filterQConstOp =
+        dyn_cast<TFL::QConstOp>(fcOp.filter().getDefiningOp());
+    auto filter = filterQConstOp.value().cast<DenseElementsAttr>();
+    auto filterShape =
+        fcOp.filter().getType().cast<RankedTensorType>().getShape();
+    std::vector<int64_t> expandedFilterVector = {1, 1, 1, 1};
```
Comment (andrewstanfordjason) [inline experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: again this is preference but I prefer:
std::vector<int64_t> expandedFilterVector = {filterShape[0], 1, 1, filterShape[1]};
I don't mind which you go with though

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    auto expandedFilterType = RankedTensorType::get(
+        expandedFilterVector,
+        fcOp.filter().getType().cast<ShapedType>().getElementType());
+    auto expandedFilterQConstOp = rewriter.create<TFL::QConstOp>(
+        fcOp.getLoc(), mlir::TypeAttr::get(expandedFilterType), filter);
+
+    // Add a Conv2D to replace the FC
+    std::vector<int64_t> expandedResultVector = {1, 1, 1, 1};
+    expandedResultVector[0] =
+        fcOp.getResult(0).getType().cast<ShapedType>().getShape()[0];
```
Comment (andrewstanfordjason) [inline experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: could fcOp.getResult(0).getType().cast<ShapedType>().getShape() have its own variable too?

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 3 above)
Comment (andrewstanfordjason) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp:69]: for all failures

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 1 above)
Comment (ghost) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: Ideally, Tablegen should be used to do this.

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 3 above)
Comment (ghost) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp:69]: I am not sure if it makes sense to use emitOpError() for match fails. 
It should be for fails which is an op error such as when the bias is too large.
What do you think?

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 2 above)
Comment (ghost) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: Will do

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 4 above)
Comment (ghost) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: Will do

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 5 above)
Comment (ghost) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp]: Will do

File: experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp (hunk 3 above)
Comment (andrewstanfordjason) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp:69]: I think you are right, this is a pattern match fail which isn't error

Comment (andrewstanfordjason) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp:69]: maybe

  /// Emit a remark about this operation, reporting up to any diagnostic
  /// handlers that may be listening.
  InFlightDiagnostic emitRemark(const Twine &message = {});


Would be a good option

Comment (ghost) [reply experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp:69]: That's a good find. I don't think we should emit it for pattern match fails though. Though I'm not sure where exactly we would use warning/remark/error.

**Review by andrewstanfordjason — APPROVED** [review]

**Pull request merged**